File: top-interview-questions-medium/Sorting and Searching/search-for-a-range/0.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
  const len = nums.length
  if (!len || nums[0] > target || nums[len - 1] < target) return [-1, -1]

  let lb = -1
  let rb = -1

  let start
  let end

  start = 0
  end = len - 1
  while (start + 1 < end) {
    let mid = Math.floor((start + end) / 2)
    if (nums[mid] < target) {
      start = mid
    } else {
      end = mid
    }
  }
  if (nums[start] === target) {
    lb = start
  } else if (nums[end] === target) {
    lb = end
  }

  start = 0
  end = len - 1
  while (start + 1 < end) {
    let mid = Math.floor((start + end) / 2)
    if (nums[mid] > target) {
      end = mid
    } else {
      start = mid
    }
  }
  if (nums[end] === target) {
    rb = end
  } else if (nums[start] === target) {
    rb = start
  }

  return [lb ,rb]
}

File: top-interview-questions-medium/Sorting and Searching/search-for-a-range/0.py
class Solution:
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        length = len(nums)
        if not length or nums[0] > target or nums[length - 1] < target: return [-1, -1]

        lb, rb = -1, -1

        start, end = 0, length - 1
        while start + 1 < end:
            mid = int((start + end) / 2)
            if nums[mid] < target:
                start = mid
            else:
                end = mid
        if nums[start] == target:
            lb = start
        elif nums[end] == target:
            lb = end

        start, end = 0, length - 1
        while start + 1 < end:
            mid = int((start + end) / 2)
            if nums[mid] > target:
                end = mid
            else:
                start = mid
        if nums[end] == target:
            rb = end
        elif nums[start] == target:
            rb = start

        return [lb, rb]
